src/main/java/org/snomed/snowstorm/core/data/services/CodeSystemUpgradeService.java

```diff
@@ -50,24 +57,89 @@ public class CodeSystemUpgradeService {
 	@Value("${snowstorm.rest-api.readonly}")
 	private boolean isReadOnly;
 
+	private static final Map<String, CodeSystemUpgradeJob> upgradeJobMap = new HashMap<>();
+
 	private final Logger logger = LoggerFactory.getLogger(getClass());
 
+	public CodeSystemUpgradeService() {
+		crunchifyTimer = new Timer();
+
+		// Schedules the specified task for repeated fixed-delay execution
+		crunchifyTimer.schedule(new CrunchifyReminder(), 0, ONE_HOUR_IN_MILLI_SEC);
+	}
+
+	public String findRunningJob(String codeSystemShortname, Integer newDependantVersion) {
+		for(String key : upgradeJobMap.keySet()) {
+			CodeSystemUpgradeJob job = upgradeJobMap.get(key);
+			if (job.getCodeSystemShortname().equalsIgnoreCase(codeSystemShortname)
+				&& job.getNewDependantVersion().compareTo(newDependantVersion) == 0
+				&& CodeSystemUpgradeJob.UpgradeStatus.RUNNING.equals(job.getStatus())) {
+				return key;
+			}
+		}
+		return null;
+	}
+
+	public String createJob(String codeSystemShortname, Integer newDependantVersion) {
+		String id = UUID.randomUUID().toString();
+		upgradeJobMap.put(id, new CodeSystemUpgradeJob(codeSystemShortname, newDependantVersion));
+		return id;
+	}
+
+	public CodeSystemUpgradeJob getJob(String jobId) {
+		return upgradeJobMap.get(jobId);
+	}
+
+	public CodeSystemUpgradeJob getUpgradeJobOrThrow(String jobId) {
+		CodeSystemUpgradeJob job = getJob(jobId);
+		if (job == null) {
+			throw new NotFoundException("Upgrade job not found.");
+		}
+		return job;
+	}
+
+	@PreAuthorize("hasPermission('ADMIN', #codeSystem.branchPath)")
+	@Async
+	public void upgradeAsync(String id, CodeSystem codeSystem, Integer newDependantVersion, boolean contentAutomations, SecurityContext securityContext) throws  ServiceException {
+		SecurityContextHolder.setContext(securityContext);
+		upgrade(id, codeSystem, newDependantVersion, contentAutomations);
+	}
+
 	@PreAuthorize("hasPermission('ADMIN', #codeSystem.branchPath)")
-	public void upgrade(CodeSystem codeSystem, Integer newDependantVersion, boolean contentAutomations) throws  ServiceException {
+	public void upgrade(String id, CodeSystem codeSystem, Integer newDependantVersion, boolean contentAutomations) throws  ServiceException {
```

Comment:
HashMap is not thread safe, but even if you were using ConcurrentHashMap, we've still got the problem of two calls coming in here at the same time even before they look at that map.    Now that you're working asynchronously, I think this method should return quickly enough that you don't need to worry about it blocking, so I think just a "synchronized" keyword in the definition of this function should protect us from race conditions.   Do you agree @CoderMChu ?

Reply:
It is thread safe to add "synchronized" to the method. However we should try to avoid using "synchronized" keyword if possible at the method level. The synchronised block should be very small to avoid blocking. It depends on the use case I think. If we only allow one upgrade at a given time then adding "synchronized" will work. But if we want to allow upgrading two different CodeSystems concurrently then this will perform poorly.